src/lib/merkle_mask/maskable_merkle_tree.ml

```diff
@@ -142,14 +145,44 @@ module Make (Inputs : Inputs_intf) = struct
     Uuid.Table.add_multi registered_masks ~key:(get_uuid t) ~data:attached_mask ;
     attached_mask
 
-  let unregister_mask_exn (t : t) (mask : Mask.Attached.t) =
+  (* TODO We have the parent already from mask.parent. remove extra param. *)
+  let rec unregister_mask_exn ?(grandchildren = `Check) (t : t)
+      (mask : Mask.Attached.t) : Mask.unattached =
     let t_uuid = get_uuid t in
     let error_msg suffix =
       sprintf "Couldn't unregister mask with UUID %s from parent %s, %s"
         (Mask.Attached.get_uuid mask |> Uuid.to_string_hum)
         (get_uuid t |> Uuid.to_string_hum)
         suffix
     in
+    [%test_result: Uuid.t]
+      ~message:"parent in param should be parent in data structure"
+      ~expect:t_uuid
+      (Mask.Attached.get_parent mask |> get_uuid) ;
+    ( match grandchildren with
+    | `Check -> (
+      match Hashtbl.find registered_masks (Mask.Attached.get_uuid mask) with
+      | Some children ->
+          failwith @@ error_msg
+          @@ sprintf
+               !"it has children that must be unregistered first: %{sexp: \
+                 Uuid.t list}"
+               (List.map ~f:Mask.Attached.get_uuid children)
+      | None ->
+          () )
+    | `I_promise_I_am_reparenting_this ->
```

Comment:
maybe "this" should be "descendants", to be clear ?

Reply:
You pass it when you're unregistering a mask with a plan to immediately register it again with a different parent. Not when you are planning to reparent any children of the mask. I suppose both situations are possible, but the latter can be accomplished by reparenting the children first without having to shut off any checks.

Reply:
I see, you're using that when reparenting the children of the original mask.

OK, can I ask for "this_mask" instead of "this" ?